lib/api/api.js

```diff
@@ -184,14 +186,103 @@ const api = {
             }
             return { returnTagCount, isImplicitDeny };
         }
+        let formDataParser;
+        let fileEventData = null;
+
+        if (apiMethod === 'objectPost') {
+            formDataParser = busboy({ headers: request.headers });
+        }
 
         return async.waterfall([
+            next => {
+                if (apiMethod === 'objectPut' || apiMethod === 'objectPutPart') {
+                    return next(null);
+                }
+                if (apiMethod === 'objectPost' && request.headers['content-type'].includes('multipart/form-data')) {
```

Comment:
If content-type doesn't crash on busboy, like `application/x-www-form-urlencoded`, it crashes later
 ```
 TypeError: Cannot read properties of undefined (reading 'key')
    at Object.objectPost (/home/mickael/scality/cloudserver/lib/api/objectPost.js:47:37)
    at /home/mickael/scality/cloudserver/lib/api/api.js:345:39
    at /home/mickael/scality/cloudserver/node_modules/async/dist/async.js:421:16
    at next (/home/mickael/scality/cloudserver/node_modules/async/dist/async.js:5302:29)
    at /home/mickael/scality/cloudserver/node_modules/async/dist/async.js:906:16
    at /home/mickael/scality/cloudserver/lib/api/api.js:310:32
    at tagConditionKeyAuth (/home/mickael/scality/cloudserver/lib/api/apiUtils/authorization/tagConditionKeys.js:78:16)
    at /home/mickael/scality/cloudserver/lib/api/api.js:301:74
    at nextTask (/home/mickael/scality/cloudserver/node_modules/async/dist/async.js:5297:14)
    at next (/home/mickael/scality/cloudserver/node_modules/async/dist/async.js:5304:9)
 ````
 AWS returns a `412 Precondition Failed` when content type is not `multipart/form-data`

Reply:
Handled here

https://github.com/scality/cloudserver/pull/5601/files#diff-69284c3576c9c4651c9ab95928cbb676d75009ef4ee5becb238b45e22bb20efeR143

Reply:
Still a lot of crashes

=> https://github.com/scality/cloudserver/pull/5601#discussion_r1670556399

> This crash if `content-type` is not defined in request
> 
> ```
> TypeError: Cannot read properties of undefined (reading 'includes')
> ```
> 
> Also it still crash if `Content-Type: a multipart/form-data`
> 
> Also with invalid/missing boundary aws shows different message than busboy does

> [!NOTE]
> It crashes if `Content-Type` is not provided:
`TypeError: Cannot read properties of undefined (reading 'includes')`
> It crashes if `Content-Type` is `abc multipart/form-data`: 
`RangeError [ERR_HTTP_INVALID_STATUS_CODE]: Invalid status code: undefined`
> It crashes if `Content-Type` is `multipart/form-data`: 
`RangeError [ERR_HTTP_INVALID_STATUS_CODE]: Invalid status code: undefined`
> => https://github.com/scality/cloudserver/pull/5601#discussion_r1670007823 and https://github.com/scality/Arsenal/pull/2248#discussion_r1669401971
> => you should not propagate busboy errors to arsenal you need to catch and throw arsenal errors

Reply:
> => you should not propagate busboy errors to arsenal you need to catch and throw arsenal errors

I feel like I'm not understanding this; is this not what you mean [here for example](https://github.com/scality/cloudserver/pull/5601/files#diff-69284c3576c9c4651c9ab95928cbb676d75009ef4ee5becb238b45e22bb20efeR122)?

Or do you mean catching at the instanciation of busboy, something like this?

```
async function parseFormData(request, response, requestContexts, log) {
    /* eslint-disable no-param-reassign */
    let formDataParser;
    if (!request.headers || !request.headers['content-type'] || !request.headers['content-type'].startsWith('multipart/form-data')) {
        return Promise.reject(errors.PreconditionFailed
            .customizeDescription('Bucket POST must be of the enclosure-type multipart/form-data'));
    }
    try {
        formDataParser = busboy({ headers: request.headers });
    } catch (err) {
        log.trace('Error creating form data parser', { error: err.toString() });
        return Promise.reject(errors.InternalError);
    }
    writeContinue(request, response);
```

I also don't understand how these got through with the functional tests I added for this...  
[Invalid content-type test
](https://github.com/scality/cloudserver/pull/5601/files#diff-406efa736d8942ad5f5ba3c3404803f3748ef0d0cd44a20b48f5ce074c3105cdR492)
[Missing content-type test](https://github.com/scality/cloudserver/pull/5601/files#diff-406efa736d8942ad5f5ba3c3404803f3748ef0d0cd44a20b48f5ce074c3105cdR540)

Is there something wrong with these tests?
How did you execute your tests to get these results?

Reply:
New commit:
cb7b4feb97ff4109390e60dfa118098422ee291f

I have removed the individual header check and instead opted for doing an egenral initialisation check that results in 412 error on failure. This covers all the tests I had previously + 2 I added for the 'abc multipart/form-data' and 'multipart/from-data abc' cases.
The tests have also been updated using classic `http` to make the request instead of `axios`

Reply:
you can remove the `require('axios')`